File: examples/36-proc-sys-and-the-agent/go/main.go
// sysagent v0 (chapter 36: /proc, /sys, and the agent) — a USE-method
// metrics collector reading /proc + /sys/fs/cgroup, no root required.
//
//	sysagent sample [--json] [--interval-ms N]
//	sysagent serve  --port P [--interval-ms N]
//
// `sample` takes one snapshot: CPU utilization from a /proc/stat delta
// spanning --interval-ms (default 200), run-queue + load from
// /proc/loadavg, memory from /proc/meminfo, per-disk I/O from
// /proc/diskstats, network rx/tx from /proc/net/dev, and cgroup PSI
// (falling back to system-wide /proc/pressure) if the kernel exposes it.
// `serve` exposes the identical snapshot as JSON over an HTTP /metrics
// endpoint, one fresh sample per request.
//
// The field names in both the --json and /metrics output are the
// deterministic, cross-language schema documented in README.md — C++ and
// Rust sysagent emit byte-for-byte the same keys.
package main

import (
	"fmt"
	"os"
	"strconv"
)

func usage() {
	fmt.Fprintln(os.Stderr,
		"usage: sysagent sample [--json] [--interval-ms N] | serve --port P [--interval-ms N]")
}

// flags is the minimal hand-rolled scan shared in shape with the C++ and
// Rust argument handling, so the three stay directly comparable.
type flags struct {
	json       bool
	intervalMs int
	port       int
}

func parseFlags(args []string) (flags, bool) {
	f := flags{intervalMs: 200, port: -1}
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--json":
			f.json = true
		case "--interval-ms":
			if i+1 >= len(args) {
				return f, false
			}
			i++
			v, err := strconv.Atoi(args[i])
			if err != nil {
				return f, false
			}
			f.intervalMs = v
		case "--port":
			if i+1 >= len(args) {
				return f, false
			}
			i++
			v, err := strconv.Atoi(args[i])
			if err != nil {
				return f, false
			}
			f.port = v
		default:
			return f, false
		}
	}
	return f, true
}

func cmdSample(args []string) int {
	f, ok := parseFlags(args)
	if !ok || f.intervalMs <= 0 {
		usage()
		return 2
	}
	snap, err := TakeSnapshot(f.intervalMs)
	if err != nil {
		fmt.Fprintln(os.Stderr, "sysagent: error:", err)
		return 1
	}
	if f.json {
		fmt.Println(ToJSON(snap))
	} else {
		fmt.Print(ToText(snap))
	}
	return 0
}

func cmdServe(args []string) int {
	f, ok := parseFlags(args)
	if !ok || f.port <= 0 || f.intervalMs <= 0 {
		usage()
		return 2
	}
	return Serve(f.port, f.intervalMs)
}

func main() {
	if len(os.Args) < 2 {
		usage()
		os.Exit(2)
	}
	var code int
	switch os.Args[1] {
	case "sample":
		code = cmdSample(os.Args[2:])
	case "serve":
		code = cmdServe(os.Args[2:])
	default:
		usage()
		code = 2
	}
	os.Exit(code)
}

File: examples/36-proc-sys-and-the-agent/rust/src/main.rs
// sysagent v0 (chapter 36: /proc, /sys, and the agent) — a USE-method
// metrics collector reading /proc + /sys/fs/cgroup, no root required.
//
//   sysagent sample [--json] [--interval-ms N]
//   sysagent serve  --port P [--interval-ms N]
//
// `sample` takes one snapshot: CPU utilization from a /proc/stat delta
// spanning --interval-ms (default 200), run-queue + load from
// /proc/loadavg, memory from /proc/meminfo, per-disk I/O from
// /proc/diskstats, network rx/tx from /proc/net/dev, and cgroup PSI
// (falling back to system-wide /proc/pressure) if the kernel exposes it.
// `serve` exposes the identical snapshot as JSON over a hand-rolled
// HTTP/1.1 /metrics endpoint, one fresh sample per request.
//
// The field names in both the --json and /metrics output are the
// deterministic, cross-language schema documented in README.md — C++ and
// Go sysagent emit byte-for-byte the same keys.
mod httpd;
mod procfs;

use std::process::ExitCode;

fn usage() {
    eprintln!(
        "usage: sysagent sample [--json] [--interval-ms N] | serve --port P [--interval-ms N]"
    );
}

/// Minimal hand-rolled flag scan, kept in the same shape as the C++ and Go
/// argument handling so the three stay directly comparable.
struct Flags {
    json: bool,
    interval_ms: u64,
    port: i64,
}

fn parse_flags(args: &[String]) -> Option<Flags> {
    let mut f = Flags {
        json: false,
        interval_ms: 200,
        port: -1,
    };
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--json" => f.json = true,
            "--interval-ms" => {
                i += 1;
                f.interval_ms = args.get(i)?.parse().ok()?;
            }
            "--port" => {
                i += 1;
                f.port = args.get(i)?.parse().ok()?;
            }
            _ => return None,
        }
        i += 1;
    }
    Some(f)
}

fn cmd_sample(args: &[String]) -> u8 {
    let Some(f) = parse_flags(args).filter(|f| f.interval_ms > 0) else {
        usage();
        return 2;
    };
    match procfs::take_snapshot(f.interval_ms) {
        Ok(snap) => {
            if f.json {
                println!("{}", procfs::to_json(&snap));
            } else {
                print!("{}", procfs::to_text(&snap));
            }
            0
        }
        Err(e) => {
            eprintln!("sysagent: error: {e}");
            1
        }
    }
}

fn cmd_serve(args: &[String]) -> u8 {
    let Some(f) = parse_flags(args).filter(|f| f.port > 0 && f.interval_ms > 0) else {
        usage();
        return 2;
    };
    httpd::serve(f.port as u16, f.interval_ms) as u8
}

fn main() -> ExitCode {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        usage();
        return ExitCode::from(2);
    }
    let code = match args[1].as_str() {
        "sample" => cmd_sample(&args[2..]),
        "serve" => cmd_serve(&args[2..]),
        _ => {
            usage();
            2
        }
    };
    ExitCode::from(code)
}
